### DISCUSSION TITLE
Extend C# "with" syntax (from C# 9 records) to work also with "structs"

### DISCUSSION THREAD

**Discussion by DrPepperBianco** [dotnet/csharplang#4350] [Language Ideas] [answered]
C# 9 added the new with syntax to copy a struct and set new values to it.

Like this:
```c#
record Person(string Name, DateTime Birthday);

Person first = new("FirstBorn", new DateTime(2021, 1,1));
Person twin = person with { Name = "SecondBorn" };
```

But I realized that this doesn't work with structs.
```c#
struct Person
{
    public string Name { get; init; }
    public DateTime Birthday { get; init; }
}

Person first = new() { Name = "FirstBorn", Birthday = new DateTime(2021, 1, 1) };
// ↓ Doesn't work!
// Person twin = first with { Name = "SecondBorn" }
```

As far as I understand this, the reason is the missing, hidden "clone" method. But for structs the clone method is not necessary because structs copy all members by default.

So as I see it, there is now reason to not let it behave the same for structs.

**Comment by ufcpp** [comment, marked as answer]
https://github.com/dotnet/csharplang/blob/master/proposals/record-structs.md#allow-with-expression-on-structs

**Reply by DrPepperBianco** [reply]
Thanks, looks interesting.
This proposal, if I understand correctly, would add a new "struct type" record struct. Then the with keyword would still not work for plain old structs. Did I understand that correctly?

I think we could discuss "with" for normal structs independently of the idea of record struct.

(That doesn't mean, that I don't like the idea of record structs. They sound interesting. But I I think on some existing code, it is not a good idea to have to change the structs just to use the "with" keyword, if it could just work with what we have now.)

**Reply by theunrepentantgeek** [reply]
I believe the intention is to enable **with** syntax to work with non-records, both classes and structs. 

Unfortunately, it's difficult to search for - as the word "with" occurs in a lot of places.

I did find a couple of mentions in LDM meeting notes however. 

On [2020-06-22](https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-06-22.md#with-expressions-on-non-records): 
> Compatibility for Clone in user-defined classes is the most difficult of these items because it relies on a new language feature for requiring that a new object is returned from the Clone method.

On [2020-09-30](https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-09-30.md):

> We have plans to enable **with** as a general pattern that any class can implement through some mechanism ...

**Reply by ufcpp** [reply]
Aside from record struct proposal, `with` expressions are extended to normal structs.

https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-09-30.md#with-expressions

> We considered whether all structs should be copyable via a with expression

**Reply by saint4eva** [reply]
> Aside from record struct proposal, `with` expressions are extended to normal structs.
> 
> https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-09-30.md#with-expressions
> 
> > We considered whether all structs should be copyable via a with expression

So, what would be the essence of record struct if the team is planning to extend "with" to non-record struct? Why not just make "with" feature available to class and struct instead of "record class" and "record struct"?

**Reply by 333fred** [reply]
> So, what would be the essence of record struct if the team is planning to extend "with" to non-record struct?

Take a look at the table in that link to see the additions that record structs bring over regular structs. 

> Why not just make "with" feature available to class and struct instead of "record class" and "record struct"?

The links that theunrepentantgeek shared detailed that we plan to do exactly that.
